code/__HELPERS/areas.dm

```diff
@@ -0,0 +1,93 @@
+#define BP_MAX_ROOM_SIZE 300
+
+// Gets an atmos isolated contained space
+// Returns an associative list of turf|dirs pairs
+// The dirs are connected turfs in the same space
+// break_if_found is a typecache of turf types to return false if found
+/proc/detect_room(turf/origin, list/break_if_found)
+	if(origin.blocks_air)
+		return list(origin)
+
+	. = list()
+	var/list/checked_turfs = list()
+	var/list/found_turfs = list(origin)
+	while(found_turfs.len)
```

Comment:
shouldn't you check that you've processed BP_MAX_ROOM_SIZE somewhere in this loop to stop it lagging the entire server?